srtcore/buffer_tools.cpp

```diff
@@ -273,5 +273,64 @@ int CSndRateEstimator::incSampleIdx(int val, int inc) const
     return val;
 }
 
+CMovingRateEstimator::CMovingRateEstimator()
+    : m_tsFirstSampleTime(sync::steady_clock::now())
+    , m_iFirstSampleIdx(0)
+    , m_iCurSampleIdx(0)
+    , m_iRateBps(0)
+    , m_Samples(NUM_PERIODS)
+{
+    resetRate(0, NUM_PERIODS);
 }
 
+void CMovingRateEstimator::addSample(int pkts, double bytes)
+{
+    const time_point now             = steady_clock::now();
+    const int        iSampleDeltaIdx = (int)count_milliseconds(now - lastSlotTimestamp) / SAMPLE_DURATION_MS;
+
+    if (iSampleDeltaIdx == 0)
+    {
+        m_Samples[m_iCurSampleIdx].m_iBytesCount += bytes;
+        m_Samples[m_iCurSampleIdx].m_iPktsCount += pkts;
+    }
+    else
+    {
+        if ((m_iCurSampleIdx + iSampleDeltaIdx) < NUM_PERIODS)
+            resetRate(m_iCurSampleIdx + 1, m_iCurSampleIdx + iSampleDeltaIdx);
+        else
+        {
+            int loopbackDiff = m_iCurSampleIdx + iSampleDeltaIdx - NUM_PERIODS;
+            resetRate(m_iCurSampleIdx + 1, NUM_PERIODS);
+            resetRate(0, loopbackDiff);
+        }
+
+        m_iCurSampleIdx                          = ((m_iCurSampleIdx + iSampleDeltaIdx) % NUM_PERIODS);
+        m_Samples[m_iCurSampleIdx].m_iBytesCount = bytes;
+        m_Samples[m_iCurSampleIdx].m_iPktsCount  = pkts;
+
+        lastSlotTimestamp = now;
+    }
+
+    computeAverageValue();
```

Comment:
As an optimization `computeAverageValue()` can be called in the `else` statement above (`if (iSampleDeltaIdx != 0)`) because you get a new period. In other cases `if (iSampleDeltaIdx == 0)` just add (bytes + pkts * HDS=R_SIZE)` to m_iRateBps. This would work with a total interval of 1 s. In a more general case has to be divided on the interval value (with integer rounding in mind).

Reply:
You're right. I though I have something I forgot about the current slot in the rate compute. This will solve this :).